### PR TITLE
[MRG+1] Traceback message in Gallery output

### PR DESCRIPTION
**Pull request by Titan-C**
This PR need comments. 
First there is the situation of how to deal with a broken build #45, discussion shall continue there.
Here I address, for the moment, that the traceback of a broken example should be also printed in the HTML output of the gallery following my comment on https://github.com/sphinx-gallery/sphinx-gallery/issues/65#issuecomment-151472930

### CODE REVIEW

**Comment by agramfort** [comment]
that's fine with me

other option is to stop the doc build with a meaningful crash message

**Comment by GaelVaroquaux** [comment]
> other option is to stop the doc build with a meaningful crash message

It should be controlled by an option (just like the building of the
gallery is).

**Comment by lesteve** [comment]
>> other option is to stop the doc build with a meaningful crash message

> It should be controlled by an option (just like the building of the gallery is).

My 2 cents: I think the current behaviour is fine for now. The general use case is to build all the examples to find all the broken ones rather than fail on the first error.

Something I would actually find useful (not sure how easy this is): have a special image for broken examples allowing them to identify quite easily in the gallery.

File: sphinx_gallery/gen_rst.py [hunk 1]
```diff
@@ -447,12 +447,14 @@ def execute_script(code_block, example_globals, image_path, fig_count,
...
-        figure_list = []
-        image_list = '%s is not compiling:' % src_file
         print(80 * '_')
-        print(image_list)
+        print('%s is not compiling:' % src_file)
         traceback.print_exc()
         print(80 * '_')
+
+        figure_list = []
+        exception_msg = indent(traceback.format_exc(), '    ')
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: Nitpick: in order not to do traceback.print_exc() and then traceback.format_exc() which is kind of doing the same thing twice, you can do:

```python

formatted_exception = traceback.format_exc()

print(formatted_exception)

exception_msg = indent(formatted_exception, ' ' * 4)
```

**Comment by lesteve** [comment]
Just tried this out and here is a snapshot:

sphinx_gallery_traceback

It would be nice to have a fixed-width font in the HTML traceback. This makes a huge difference for long tracebacks.

**Comment by Titan-C** [comment]
Now it's improved.
I painted the broken image that takes over in the gallery thumbnails on failed examples. And the traceback message is not anymore in the sphinx warning box that changes font style but uses just a container.

**Comment by lesteve** [comment]
> Now it's improved.
I painted the broken image that takes over in the gallery thumbnails on failed examples. And the traceback message is not anymore in the sphinx warning box that changes font style but uses just a container.

The image thing is quite nice, although we can think about tweaking the "broken" image.

I still get non fixed-width font in the warnings when generating the gallery locally though.

**Comment by Titan-C** [comment]
> I still get non fixed-width font in the warnings when generating the gallery locally though.

No idea here. What does you browser tell you the CSS style is. The block shall be a sphx-glr-traceback. That one has no style for fonts. The example I'm providing gives me this

snapshot1

**Comment by lesteve** [comment]
I have to say not sure whether the technical term for it is fixed-font or not but basically what I was trying to say is that I would have the same font in the traceback and in the code cell.

**Comment by Titan-C** [comment]
> I would have the same font in the traceback and in the code cell.

I just learned that is syntax hightlighting for python traceback messages. Thus I'm making the traceback outputs be parsed as that. Plus an additional light red box to call attention is a traceback and that the example fails.

image

**Comment by Titan-C** [comment]
As here I change some CSS I would like PR #80 to be merged first. This branch is already rebased to it, but it did not generate a branch dependency as I expected.

**Comment by agramfort** [comment]
is it possible to have the build stop if an example fails? to me that should be the default behavior. It happened to me too many times to push a website update with broken examples... it happens quickly when you have dozens of examples.

**Comment by Titan-C** [comment]
I added the quick fail option to break the gallery build process as soon as an example trows an Exception, added the documentation what to put in the makefile for that. But I have not made it the default behavior, as building the complete gallery and scanning for the broken examples visually from the rendered website seems more intuitive for me. You can vote on that.

File: doc/advanced_configuration.rst [hunk 2]
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
 save it to the current folder. This is a striped version of the
 Sphinx-Gallery module to incorporate in your project. You should also
 add it to your version control system.
+
+Build the Gallery without executing the examples
+================================================
+
+Sphinx-Gallery can parse all your examples and build the gallery
+without executing any of the scripts. This is just for speed
+visiulization processes of the gallery and the size it take your
```
Comment (agramfort) [inline doc/advanced_configuration.rst]: visiulization

Comment (agramfort) [inline doc/advanced_configuration.rst]: take -> takes

display -> to display

File: doc/advanced_configuration.rst [hunk 3]
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
...
+website display, or any use you can imagine for it. To achive this you
```
Comment (agramfort) [inline doc/advanced_configuration.rst]: achive -> achieve

File: doc/advanced_configuration.rst [hunk 4]
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
...
+
+Remember that for ``Makefile`` white space is significant and the indentation are tabs
+and not spaces
+
+Dealing with failing Gallery example scripts
+============================================
+
+As your project evolves some of your example scripts might stop
+executing properly. Sphinx-Gallery assist you in the discovery process
+of those bugged examples. The defaul behavior is to replace the
```
Comment (agramfort) [inline doc/advanced_configuration.rst]: defaul -> default

File: doc/advanced_configuration.rst [hunk 5]
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
...
+behavior.
+
+An extra functionality of Sphinx-Gallery is the early fail option. In
+this mode the gallery build process breaks as soon as an exeption
+occurs in the execution of the examples scripts. To activate this
+modality you need to pass a flag at the build process. It can be done
+by including in your ``Makefile``::
+
+    html-quick_fail:
+        $(SPHINXBUILD) -D quick_fail=1 -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
```
Comment (agramfort) [inline doc/advanced_configuration.rst]: quick_fail -> fail_on_error ?

**Comment by agramfort** [comment]
besides LGTM

File: doc/advanced_configuration.rst (hunk 5 above)
Comment (Titan-C) [inline doc/advanced_configuration.rst]: This one I was also thinking, early_fail, early_abort

Comment (agramfort) [inline doc/advanced_configuration.rst]: > This one I was also thinking, early_fail, early_abort

I find fail_on_error more explicit.

File: sphinx_gallery/gen_rst.py [hunk 6]
```diff
@@ -438,21 +438,35 @@ def execute_script(code_block, example_globals, image_path, fig_count,
...
+
         print(80 * '_')
-        print(image_list)
-        traceback.print_exc()
+        print('%s is not compiling:' % src_file)
+        print(formatted_exception)
         print(80 * '_')
+
+        figure_list = []
+        image_list = codestr2rst(formatted_exception, 'pytb')
```
Comment (GaelVaroquaux) [inline sphinx_gallery/gen_rst.py]: For readability, I would suggest being explicit about the keyword argument here, by writing:

<pre>
    ..., lang='pytb')
</pre>

**Comment by GaelVaroquaux** [comment]
I agree with your various design choices (in particular the default).

+1 for merge. I have suggested a tiny style change.

**Comment by agramfort** [comment]
it's fine to merge as is for me

thx @Titan-C

File: doc/Makefile [hunk 7]
```diff
@@ -58,6 +58,11 @@ html-noplot:
 	@echo
 	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
 
+html-quick_fail:
```
Comment (lesteve) [inline doc/Makefile]: maybe html-quick-fail in order not to mix - and _

File: examples/plot_raise.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+
+plt.pcolormesh(np.random.randn(100, 100))
+
+
+###############################################################################
+# Here is another error raising Block
+
+raise IOError
```
Comment (lesteve) [inline examples/plot_raise.py]: Maybe put a more realistic error like: `plt.plot('this is not going to work')`

Comment (lesteve) [inline examples/plot_raise.py]: What I meant is that the traceback in this case will be more interesting.

**Comment by lesteve** [comment]
Is broken_stamp.svg actually used?

Slightly related is do you know why the "broken" image png appears with a black background in the gallery:
sphinx-gallery

**Comment by lesteve** [comment]
My contribution for the 'quick_fail' naming: abort_on_example_error.

**Comment by lesteve** [comment]
Could you rebase on master so that we make sure no extra CSS changes has slipped through in this PR?

**Comment by Titan-C** [comment]
broken_stamp.svg is not used now, but is the source of for the png file. Not sure if it will become useful in the future if we want to tune the image or if #61 

The background of the image is black in the thumbnails because in the sphinx-classic css theme all body backgrounds are black, and the png image is transparent.

File: examples/plot_raise.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+"""
+=============================
+Examples that fail to execute
```
Comment (lesteve) [inline examples/plot_raise.py]: Singular seems more fitting here: Example that fails to execute

**Comment by GaelVaroquaux** [comment]
Overall, LGTM. +1 for merge (ideally, please address the comments of the others).
Thanks!

**Comment by agramfort** [comment]
fine with me

**Comment by Titan-C** [comment]
Rebased to master and comments addressed

**Comment by GaelVaroquaux** [comment]
@agramfort was good with this. I am too. Let's merge!

**Pull request merged**